docs/enhancements/image-handling-service.md

```diff
@@ -0,0 +1,155 @@
+expose all the information the image service needs to create the discovery ISO.
+Specifically, the values that are required and not currently accessible over
+the API are:
+- Pull secret
+
+### Risks and Mitigations
+
+Relying strictly on streaming will increase the load on the image service in
+the SaaS deployment as presigned links directly to S3 would no longer be provided.
+Downloading the image through the service is already required when running
+on a local cluster so this shouldn't be a significant problem, but should
+be load tested. This will be further mitigated by splitting the service as
+many downloads won't affect the other operations of assisted service.
+
+
+## Design Details [optional]
+
+
+### Open Questions
+
+- How should the image service API be defined?
+  - A separate swagger file?
+    - Will this require a separate client package?
+- How does traffic get to the new service?
+  - [Path based routes?](https://docs.openshift.com/container-platform/4.7/networking/routes/route-configuration.html#nw-path-based-routes_route-configuration?)
+  - An entirely separate route?
+  - How does this choice affect the previous point about swagger files?
+- Where does the source live?
+  - openshift/assisted-service? (Like controllers)
+  - New repository in the OpenShift GitHub organization?
+  - Elsewhere?
+- How does a user access the discovery ignition?
+  - Currently this is stored when the image is generated.
+  - Could be rendered and served on-demand from the image service.
+
+### UI Impact
+
+Allowing only streaming would mean that the discovery image download modal would
+need to change to remove the presigned S3 URL and instead point users to the
+service directly.
+
+Any existing workarounds regarding time between generating images can be removed
+when images are no longer being generated in the background.
+
+### Test Plan
+
+The image service can, and should, be tested in isolation using a mock assisted
+service API which will return some image info.
+
+As mentioned previously, load testing will also be needed on the SaaS
+deployment to ensure it can handle whatever peak image download load is
+expected.
+
+## Drawbacks
+
+- Maintaining and deploying an entirely separate service is complicated.
+- No longer saving the image makes any future requirement to serve directly to a
+  BMC much more difficult to implement.
+- Adds traffic and load to the SaaS by removing images from S3.
+
+## Alternatives
+
+This proposal could be amended to handle either streaming or storing an image,
+but it would complicate authentication and authorization.
+
+In the current proposal every call to the image service (download image, and
+maybe download discovery ignition) would require a call to assisted service to
+fetch customization information. If the image is generated and stored by the
+image service, the download call will not need a call to assisted service and
+would require some other means of authentication and authorization.
```

Comment:
The JWTs are nice because they can be generated in one place but verified by anyone with the public half of the key.

If we can eliminate a requirement for assisted-service to be up in order to serve an image, that would help by adding HA capability for image serving despite assisted-service being a singleton.

Reply:
> The JWTs are nice because they can be generated in one place but verified by anyone with the public half of the key.

I think authenticating the JWT is not the problem. The issue is really about how we currently do revocation (querying the database to see if some object exists) and if we every want to implement some form of authorization which would likely also be described in the database.

> If we can eliminate a requirement for assisted-service to be up in order to serve an image, that would help by adding HA capability for image serving despite assisted-service being a singleton.

This isn't possible as the proposal is written. We still need to query the image customization data from assisted service. This is true even if we had no authentication at all.

Reply:
Gotcha. If we were letting the assisted-service do the heavy lifting of generating ignition, and the image service was mostly there to assemble the pieces and serve them efficiently, then there could be an opportunity to cache the ignition and anything else from assisted-service. But it sounds like that may be a different design than you have in mind.